### PR TITLE
cirrus: enable cgroups v2 tests with crun

### PR DESCRIPTION
**Pull request by giuseppe**
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM
Don't see why this is a WIP.  If it passes it should be merged.

**Comment by giuseppe** [comment]
> LGTM
> Don't see why this is a WIP. If it passes it should be merged.

I've opened the PR just to see what would fail and keep working on it.  I had expected a lot of failures, as the cgroup v2 backend is not really tested.  From 70+ failures it went down now to a few ones.  There were a couple of issues in crun, but most of the failures were related to the tests not ready for cgroups v2.

**Comment by giuseppe** [comment]
:tada: :tada: :tada:  cgroups 2 tests are passing now :tada: :tada: :tada:

**Comment by giuseppe** [comment]
it is using the master branch for crun.  I'll cut a new release in the next days

**Comment by giuseppe** [comment]
tests are all green

**Comment by giuseppe** [comment]
@cevich @mheon PTAL

**Comment by mheon** [comment]
This is just test changes, right?

I don't like the duplication around determining if we're in V2 mode, but I don't know if we can remove it easily.

Otherwise LGTM

**Comment by giuseppe** [comment]
yes, it is just tests code changes beside: https://github.com/containers/libpod/pull/3796/commits/04d333f0efd0d1af78785c76ed357f0fa98cafa1

**Comment by giuseppe** [comment]
@cevich I think we will need to enable the same tests on the crun repository, so that we don't risk to break Podman.  How difficult would it be to run Podman tests there?

File: contrib/cirrus/integration_test.sh [hunk 1]
```diff
@@ -37,11 +37,16 @@ case "$SPECIALMODE" in
...
+        setenforce 0
+        yum builddep -y crun
+        (git clone --depth=1 https://github.com/containers/crun && cd crun && ./autogen.sh && ./configure --prefix=/usr && make -j4 && make install)
+        export OCI_RUNTIME=/usr/bin/crun
         make
         make install PREFIX=/usr ETCDIR=/etc
         make test-binaries
-        echo "WARNING: Integration tests not yet ready for cgroups V2"
-        #TODO: make local${TESTSUITE}
+        make local${TESTSUITE}
```
Comment (cevich) [inline contrib/cirrus/integration_test.sh]: Are you intending to support running of the system-tests?  If so, please add to `.cirrus.yml`:

```yaml
special_testing_cgroupv2_task
    ...
    system_test_script: '$SCRIPT_BASE/system_test.sh |& ${TIMESTAMP}'
```

**Comment by cevich** [comment]
> How difficult would it be to run Podman tests there?

I'm not a big fan of running "child" tests in "parent" projects, it tends to result in very messy environments (in the parent) which don't accurately test compatibility and require an extraordinary amount of maintenance and duct-tape.  

OTOH, I see that crun looks to already have some dedicated testing setup in travis?.  

If you need help converting or extend that to something more modern than old/crusty Ubuntu Xenial host environment, that's something I could help with.  It's not trivial, but having existing/functional automation helps a lot.

If it's a short/medium term automation need, I can get something going in a day or two worth of work (need a jira card).  If there's a chance it will be long-term (beyond 6-months), then I'd prefer to setup a dedicated cloud-environment (like we do for c/storage, buildah, etc.).  That adds about 4-hours of work for me but a few days waiting for e-mail requests/approvals for funding.  I'm fine either way, but would rather get started on dedicated long-term stuff early, if that's what is needed.

**Comment by giuseppe** [comment]
if it is too much effort, I can open WIP PR to containers/libpod once in a while and before cutting a new release to verify it still works

File: contrib/cirrus/integration_test.sh (hunk 1 above)
Comment (giuseppe) [reply contrib/cirrus/integration_test.sh]: I've not yet looked into them, I'd expect them to fail at this point.  We can add them later, once we are sure they work with a different OCI runtime

File: contrib/cirrus/integration_test.sh [hunk 2]
```diff
@@ -37,11 +37,16 @@ case "$SPECIALMODE" in
                 -o CheckHostIP=no $GOSRC/$SCRIPT_BASE/rootless_test.sh ${TESTSUITE}
         ;;
     cgroupv2)
+        # FIXME: use the package once all the fixes are in a release
+        # yum install -y crun
+        setenforce 0
+        yum builddep -y crun
+        (git clone --depth=1 https://github.com/containers/crun && cd crun && ./autogen.sh && ./configure --prefix=/usr && make -j4 && make install)
```
[the remark this answers is not in the document]
Comment (giuseppe) [reply contrib/cirrus/integration_test.sh]: we need to install from source for now, as master contains some fixes necessary to get the tests passing

**Comment by giuseppe** [comment]
> I'm with Matt, that's a lot of duplication. IMHO, this PR should depend on #3754 (includes crun package) rather than inline package + source install (see my comments above). It's one of my top priorities this week to get that merged ASAP - CI in other repos depend on it as well.

for now we need to use crun from git as there are fixes required for the tests.  I'll cut a new release and then we can use the package

**Comment by cevich** [comment]
> if it is too much effort, I can open WIP PR to containers/libpod once in a while and before cutting a new release to verify it still works

That would probably be okay a few times, then get to be a major pain-point, don't you think?

Since crun is required for this cgroupsv2 test, it'll get a fairly big safety net that way (it's going to run for everybody's changes and post-merge also).  It sounds like you're mainly worried about new, crun-specific changes affecting downstream compatibility, eh?

Traditionally, this would be where crun would have it's own integration and system-tests, as a kind of "contract" with the "outside world" (libpod).  So for example, if downstream changes in libpod require different crun behaviors, the tests in crun need to be modified accordingly.

Honestly there are simply not a lot of quick/cheap/easy ways to solve this, which also play nicely with (potentially many) other parallel callers/users/developers of different crun versions/releases (including libpod).  Testing and automation is just hard work sometimes.

That said, I'm willing to do most/all the automation implementation work.  All I need is you/other community members to support me if I have crun-specific issues/questions along the way. It's _way_ better if we plan and get it right/good early on.  Doing/Re-doing it later is almost always much more difficult.

Which jira cards should I write up for this (I'll CC you), short/medium term or long-term or other?

**Comment by cevich** [comment]
> for now we need to use crun from git as there are fixes required for the tests. I'll cut a new release and then we can use the package

Okay, that's fine if it's just for now.  Just please make sure there's a "post-it note" (or jira card or in-line comment) somewhere so it's actually "just for now".  I can't say how many times these have slipped and become permanent debugging-nightmares for `$FUTURE_DEVELOPER`.  Best to avoid that.

**Comment by giuseppe** [comment]
> Which jira cards should I write up for this (I'll CC you), short/medium term or long-term or other?

Fedora 31 is getting closer, so I think we should try to get at least these tests running for Podman as soon as we can, and have some sort of confidence with cgroups v2.  
We can discuss the details next week on how to coordinate and proceed with the crun integration.

**Pull request merged**